### DISCUSSION TITLE
Apache FOP to Antennahouse : how to remove illegal characters from index entry ids ?

### DISCUSSION THREAD

**Discussion by louislapicque** [dita-ot/dita-ot#4149] [Q&A]
The short version: when moving from FOP to AHF, new errors are revealed by AHF. I'm unsure how to investigate, what stylesheets may be related to the issue. 

Given the following DITA  topic :

`<topic id="topic_test">
    <title>Topic test</title>
    <prolog>
        <metadata>
            <keywords>
                <indexterm>Test category: term A</indexterm>
            </keywords>
        </metadata>
    </prolog>`
(...)

When generating the PDF using our plugin, the page number isn't printed (the value is '0', and the link doesn't work).

The following error is shown in the console:
[ahf] Error Level : 2
[ahf] Error Code  : 10761 (2A09)
[ahf] Invalid property value: id="Test category: term A:_unique_d169e28".
[ahf] Line 21, Col 1285, (temporary folder)\topic.fo

The following topic.fo is generated:
`<fo:block border-end-width="0pt" border-start-width="0pt" keep-with-next.within-column="always" line-height="100%">
<fo:marker marker-class-name="current-header">
<fo:inline xmlns:xs="http://www.w3.org/2001/XMLSchema">1 </fo:inline>Topic test</fo:marker><fo:wrapper id="unique_1"/>
<fo:wrapper id="_OPENTOPIC_TOC_PROCESSING_d169e15"/>
<fo:wrapper id="Test category: term A:_unique_d169e28"/>
<fo:inline xmlns:xs="http://www.w3.org/2001/XMLSchema">1 </fo:inline>Topic test</fo:block>`

Since @id contains the colon and space characters, it seems to me something isn't right, but I'm stumped. 
Any help, any question welcome! Many thanks.

**Comment by chrispy-snps** [comment]
@louislapicque - are you able to contact Antennahouse support to ask them? It looks like they need to normalize the element's `@id` to a valid ID value when computing it. Perhaps they could also point you to the relevant template to help you find a workaround.

**Reply by louislapicque** [reply]
Hi Chris, thanks for the suggestion.

Yes I've been in touch with them, and they were very helpful in helping to find a solution. 

Here's a takeaway (for me at least, possibly for others as well):

FOP uses **@internal-destination** and **fo:page-number-citation/@ref-id** to generate the @id.
(the @id value may contain spaces and colons which are invalid as @id characters)

AH Formatter uses the **fo:wrapper/@index-key** and the **fo:index-page-citation-list/
fo:index-key-reference/@ref-index-key** mechanism introduced in XSL 1.1.
